File: 5-Problem-A-Day/333. Largest BST Subtree.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode left; TreeNode
 * right; TreeNode(int x) { val = x; } }
 */
class Solution {
    public int largestBSTSubtree(TreeNode root) {
        int[] ans = expand(root);
        return ans[2];
    }

    private int[] expand(TreeNode root) {
        if (root == null)
            return new int[] {Integer.MAX_VALUE, Integer.MIN_VALUE, 0};
        int[] left = expand(root.left);
        int[] right = expand(root.right);
        if (root.val > left[1] && root.val < right[0])
            return new int[] {Math.min(root.val, left[0]), Math.max(root.val, right[1]),
                    left[2] + right[2] + 1};
        else
            return new int[] {Integer.MIN_VALUE, Integer.MAX_VALUE, Math.max(left[2], right[2])};
    }
}

File: 2021/333. Largest BST Subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxCount = 0;
    int largestBSTSubtree(TreeNode* root) {
        largestBSTtree(root);
        return maxCount;
    }
    
    array<int, 3> largestBSTtree(TreeNode* root) {
        if (root == NULL) return {0, 0, 0};
        auto left = largestBSTtree(root->left);
        auto right = largestBSTtree(root->right);
        if (left[2] < 0 || right[2] < 0) {
            return {0, 0, -1};
        }
        if (left[2] == 0 && right[2] == 0) {
            maxCount = max(maxCount, 1);
            return {root->val, root->val, 1};
        }        
        if (left[2] == 0) {
            if (right[0] <= root->val) {
                return {0, 0, -1};
            } else {
                maxCount = max(maxCount, right[2] + 1);
                return {root->val, right[1], right[2] + 1};
            }
        } 
        if (right[2] == 0) {
            if (left[1] >= root->val ) {
                return {0, 0, -1};
            } else {
                maxCount = max(maxCount, left[2] + 1);
                return {left[0], root->val, left[2] + 1};
            }
        }
        
        if (left[1] >= root->val || right[0] <= root->val) {
            return {0, 0, -1};
        }
        maxCount = max(maxCount, left[2] + right[2] + 1);
        return {left[0], right[1], left[2] + right[2] + 1};
        
    }
};